src/DynamoCLI/GeometryHolder.cs

```diff
@@ -0,0 +1,144 @@
+                if (!nodeModel.RequestVisualUpdateAsync(model.Scheduler, model.EngineController, factory))
+                {
+                    Done.Set();
+                }
+            }
+
+            public void Dispose()
+            {
+                node.RenderPackagesUpdated -= NodeRenderPackagesUpdated;
+            }
+
+            private void NodeRenderPackagesUpdated(NodeModel nodeModel, RenderPackageCache renderPackages)
+            {
+
+                if (renderPackages.Packages.Count() > 0)
+                {
+                    List<double[]> verts = new List<double[]>();
+                    List<List<byte>> vertColors = new List<List<byte>>();
+                    List<double[]> normals = new List<double[]>();
+                    List<List<double>> points = new List<List<double>>();
+                    List<List<byte>> pointColors = new List<List<byte>>();
+                    List<List<double>> lines = new List<List<double>>();
+                    List<List<byte>> lineColors = new List<List<byte>>();
+
+                    foreach (IRenderPackage p in renderPackages.Packages)
+                    {
+                        var meshVertices = p.MeshVertices.ToArray();
+                        if (meshVertices.Length > 0)
+                        {
+                            verts.Add(meshVertices);
+                        }
+
+                        var meshVertexColors = p.MeshVertexColors.ToList();
+                        if(meshVertexColors.Count > 0)
+                        {
+                            vertColors.Add(meshVertexColors);
+                        }
+
+                        var meshNormals = p.MeshNormals.ToArray();
+                        if (meshNormals.Length > 0)
+                        {
+                            normals.Add(meshNormals);
+                        }
+
+                        var pointVertices = p.PointVertices.ToList();
+                        if (pointVertices.Count > 0)
+                        {
+                            points.Add(pointVertices);
+                        }
+
+                        var pointVertexColors = p.PointVertexColors.ToList();
+                        if (pointVertexColors.Count > 0)
+                        {
+                            pointColors.Add(pointVertexColors);
+                        }
+
+                        var lineStripVertices = p.LineStripVertices.ToList();
+                        if (lineStripVertices.Count > 0)
+                        {
+                            lines.Add(lineStripVertices);
+                        }
+
+                        var lineStripVertexColors = p.LineStripVertexColors.ToList();
+                        if (lineStripVertexColors.Count > 0)
+                        {
+                            lineColors.Add(lineStripVertexColors);
+                        }
+                    }
+
+                    groupData.Add("name", nodeModel.GUID.ToString());
```

Comment:
hmm, couldn't this dataStructure just be a function that takes a renderPackage and serializes it?

Reply:
Is the count > 0 due to custom nodes?